File: src/Core.Python/schema_evolution_golden.py
#!/usr/bin/env python3
"""Schema evolution golden vector conformance — Python oracle (#5 of 10).

Parses schema-golden-vectors.json, replays deltas, asserts value-equality.
Zero dependencies beyond stdlib (json, sys).
"""

import json
import sys


def apply_delta(schema: list[tuple[dict, int]], delta: dict) -> list[tuple[dict, int]]:
    """Apply a schema evolution delta. Retract at -1, insert at +1, drop weight=0."""
    # Build name → (field, weight) map
    m: dict[str, tuple[dict, int]] = {}
    for field, weight in schema:
        name = field["name"]
        if name in m:
            m[name] = (field, m[name][1] + weight)
        else:
            m[name] = (field, weight)

    for field in delta["retract"]:
        name = field["name"]
        if name in m:
            m[name] = (field, m[name][1] - 1)
        else:
            m[name] = (field, -1)

    for field in delta["insert"]:
        name = field["name"]
        if name in m:
            m[name] = (field, m[name][1] + 1)
        else:
            m[name] = (field, 1)

    # Drop zero-weight entries
    return [(field, w) for field, w in m.values() if w != 0]


def active_fields(schema: list[tuple[dict, int]]) -> list[dict]:
    return [field for field, w in schema if w > 0]


def sorted_field_names(schema: list[tuple[dict, int]]) -> list[str]:
    return sorted(f["name"] for f, w in schema if w > 0)


def main() -> int:
    if len(sys.argv) < 2:
        print(
            "Usage: python schema_evolution_golden.py <path-to-json>", file=sys.stderr
        )
        return 1

    with open(sys.argv[1]) as f:
        vectors = json.load(f)

    initial_fields = vectors["initialFields"]
    deltas = vectors["deltas"]
    expected_states = vectors["expectedReplayStates"]
    expected_final = vectors["expectedFinalState"]
    comm_pairs = vectors["commutativePairs"]

    # Initialize schema
    schema = [(field, 1) for field in initial_fields]

    # Replay deltas
    print("--- Replaying deltas ---")
    replay_states = []
    for i, delta in enumerate(deltas):
        schema = apply_delta(schema, delta)
        replay_states.append(schema)

        active = active_fields(schema)
        expected = expected_states[i]

        assert len(active) == expected["entryCount"], (
            f"Delta {i} count mismatch: expected {expected['entryCount']}, got {len(active)}"
        )
        print(f"  Delta {i}: {len(active)} fields ✓")

    # Final state
    print("--- Final state ---")
    final_names = sorted_field_names(schema)
    assert final_names == expected_final["fieldNames"], (
        f"Final names mismatch: {final_names} != {expected_final['fieldNames']}"
    )
    assert len(active_fields(schema)) == expected_final["fieldCount"]
    print(f"  Final: {len(final_names)} fields [{', '.join(final_names)}] ✓")

    # Commutativity
    print("--- Commutativity ---")
    initial_schema = [(field, 1) for field in initial_fields]
    for pair in comm_pairs:
        a, b = pair["deltaA"], pair["deltaB"]
        state_ab = apply_delta(apply_delta(initial_schema, deltas[a]), deltas[b])
        state_ba = apply_delta(apply_delta(initial_schema, deltas[b]), deltas[a])

        names_ab = sorted_field_names(state_ab)
        names_ba = sorted_field_names(state_ba)
        assert names_ab == names_ba, f"Deltas ({a},{b}) do not commute"
        print(f"  Deltas ({a},{b}) commute ✓")

    print("\nAll golden vectors passed! (Python oracle #5)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/Core.Go/schema_evolution_golden.go
// Schema evolution golden vector conformance — Go oracle (#6 of 10).
// Parses schema-golden-vectors.json, replays deltas, asserts value-equality.
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
)

type SchemaField struct {
	Name     string `json:"name"`
	Type     string `json:"type"`
	Required bool   `json:"required"`
}

type SchemaEntry struct {
	Field  SchemaField
	Weight int
}

type Delta struct {
	Retract []SchemaField `json:"retract"`
	Insert  []SchemaField `json:"insert"`
}

type ReplayState struct {
	ActiveFields []SchemaField `json:"activeFields"`
	EntryCount   int           `json:"entryCount"`
}

type FinalState struct {
	FieldNames []string `json:"fieldNames"`
	FieldCount int      `json:"fieldCount"`
}

type CommPair struct {
	DeltaA   int  `json:"deltaA"`
	DeltaB   int  `json:"deltaB"`
	Commutes bool `json:"commutes"`
}

type GoldenVectors struct {
	InitialFields        []SchemaField `json:"initialFields"`
	Deltas               []Delta       `json:"deltas"`
	ExpectedReplayStates []ReplayState `json:"expectedReplayStates"`
	ExpectedFinalState   FinalState    `json:"expectedFinalState"`
	CommutativePairs     []CommPair    `json:"commutativePairs"`
}

func applyDelta(schema []SchemaEntry, delta Delta) []SchemaEntry {
	m := make(map[string]SchemaEntry)
	for _, e := range schema {
		if existing, ok := m[e.Field.Name]; ok {
			m[e.Field.Name] = SchemaEntry{Field: e.Field, Weight: existing.Weight + e.Weight}
		} else {
			m[e.Field.Name] = e
		}
	}
	for _, f := range delta.Retract {
		if existing, ok := m[f.Name]; ok {
			m[f.Name] = SchemaEntry{Field: f, Weight: existing.Weight - 1}
		} else {
			m[f.Name] = SchemaEntry{Field: f, Weight: -1}
		}
	}
	for _, f := range delta.Insert {
		if existing, ok := m[f.Name]; ok {
			m[f.Name] = SchemaEntry{Field: f, Weight: existing.Weight + 1}
		} else {
			m[f.Name] = SchemaEntry{Field: f, Weight: 1}
		}
	}
	var result []SchemaEntry
	for _, e := range m {
		if e.Weight != 0 {
			result = append(result, e)
		}
	}
	return result
}

func sortedFieldNames(schema []SchemaEntry) []string {
	var names []string
	for _, e := range schema {
		if e.Weight > 0 {
			names = append(names, e.Field.Name)
		}
	}
	sort.Strings(names)
	return names
}

func activeCount(schema []SchemaEntry) int {
	count := 0
	for _, e := range schema {
		if e.Weight > 0 {
			count++
		}
	}
	return count
}

func sliceEqual(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: go run . <path-to-json>")
		os.Exit(1)
	}

	data, err := os.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read: %v\n", err)
		os.Exit(1)
	}

	var vectors GoldenVectors
	if err := json.Unmarshal(data, &vectors); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to parse: %v\n", err)
		os.Exit(1)
	}

	// Initialize
	schema := make([]SchemaEntry, len(vectors.InitialFields))
	for i, f := range vectors.InitialFields {
		schema[i] = SchemaEntry{Field: f, Weight: 1}
	}

	// Replay
	fmt.Println("--- Replaying deltas ---")
	for i, delta := range vectors.Deltas {
		schema = applyDelta(schema, delta)
		active := activeCount(schema)
		expected := vectors.ExpectedReplayStates[i].EntryCount
		if active != expected {
			fmt.Fprintf(os.Stderr, "FAIL: Delta %d count mismatch: expected %d, got %d\n", i, expected, active)
			os.Exit(1)
		}
		fmt.Printf("  Delta %d: %d fields ✓\n", i, active)
	}

	// Final
	fmt.Println("--- Final state ---")
	finalNames := sortedFieldNames(schema)
	if !sliceEqual(finalNames, vectors.ExpectedFinalState.FieldNames) {
		fmt.Fprintf(os.Stderr, "FAIL: Final names mismatch\n")
		os.Exit(1)
	}
	if activeCount(schema) != vectors.ExpectedFinalState.FieldCount {
		fmt.Fprintf(os.Stderr, "FAIL: Final count mismatch\n")
		os.Exit(1)
	}
	fmt.Printf("  Final: %d fields [%s] ✓\n", len(finalNames), join(finalNames))

	// Commutativity
	fmt.Println("--- Commutativity ---")
	initialSchema := make([]SchemaEntry, len(vectors.InitialFields))
	for i, f := range vectors.InitialFields {
		initialSchema[i] = SchemaEntry{Field: f, Weight: 1}
	}
	for _, pair := range vectors.CommutativePairs {
		ab := applyDelta(applyDelta(initialSchema, vectors.Deltas[pair.DeltaA]), vectors.Deltas[pair.DeltaB])
		ba := applyDelta(applyDelta(initialSchema, vectors.Deltas[pair.DeltaB]), vectors.Deltas[pair.DeltaA])
		if !sliceEqual(sortedFieldNames(ab), sortedFieldNames(ba)) {
			fmt.Fprintf(os.Stderr, "FAIL: Deltas (%d,%d) do not commute\n", pair.DeltaA, pair.DeltaB)
			os.Exit(1)
		}
		fmt.Printf("  Deltas (%d,%d) commute ✓\n", pair.DeltaA, pair.DeltaB)
	}

	fmt.Println("\nAll golden vectors passed! (Go oracle #6)")
}

func join(s []string) string {
	result := ""
	for i, v := range s {
		if i > 0 {
			result += ", "
		}
		result += v
	}
	return result
}
